### PR TITLE
Basic streaming

### PR DESCRIPTION
**Pull request by phischu**
This one would greatly benefit from #252.

### CODE REVIEW

File: libraries/common/stream.effekt
Comment (marzipankaiser) [inline libraries/common/stream.effekt:16]: I'd want the program to either panic or act as if it is 0. (This is not something where I'd expect forever, and this is a place where I could see myself accidentally passing a negative number).

File: libraries/common/stream.effekt
Comment (marzipankaiser) [inline libraries/common/stream.effekt:135]: I usually call this function either `sum` or `summing` (to distinguish it from `sum(List[Int]): Int` etc).

File: libraries/common/stream.effekt
Comment (marzipankaiser) [inline libraries/common/stream.effekt:156]: I would have expected this to do the standard "build a reversed list then reverse" trick for me, so it does not build up stack in the handler.
That is, if this is actually helpful in our current implementation.

File: libraries/common/stream.effekt
Comment (b-studios) [reply libraries/common/stream.effekt:16]: Yes, please check and panic

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:74]: Would be nice to see a few examples to come up with a less punny name.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:77]: I like indexed somehow (even though it introduces boxing).

Anyway, either we use it everywhere (including list, array, etc.) or nowhere. But then indexed should be moved to `effekt.effekt`.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:79]: Same for this, this should be in `effekt.effekt` somewhere in the toplevel documentation.

**Comment by jiribenes** [comment]
It would be really nice to have a couple of examples showcasing the various features so that we can get an idea of how to use it :)

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:88]: While not call this `until`? Then the usage would look like:

```
for { until(10) } { ... }
```

Comment (b-studios) [reply libraries/common/stream.effekt:88]: hm... `rangeTo` does make sense now I think more about it.

We could even use it as

```
for { rangeTo(1); rangeTo(2); rangeTo(3) } { ... }
```
which streams

1, 1, 2, 1, 2, 3

And reads better than `until`.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:114]: Common....this is waaaay to punny. Like `ret.urn`...

**Comment by jiribenes** [comment]
By the way, here's Unison's [`Stream` effect](https://share.unison-lang.org/@unison/base/code/main/latest/types/@rfi1v9429f9qluv533l2iba77aadttilrpmnhljfapfnfa6sru2nr8ibpqvib9nc4s4nb9s1as45upsfqfqe6ivqi2p82b2vd866it8)  and [`Each` effect](https://share.unison-lang.org/@unison/base/code/releases/3.13.0/latest/types/abilities/Each) (lazy nondeterminism from a stream)

```hs
Each.toList do
  use Nat <
  a = Each.range 0 5
  b = each [1, 2, 3]
  guard (a < b)
  (a, b)
```

File: libraries/common/stream.effekt
Comment (b-studios) [reply libraries/common/stream.effekt:114]: Maybe `always`?

File: libraries/common/stream.effekt
Comment (jiribenes) [reply libraries/common/stream.effekt:135]: I've posted the related work in Unison below :)
https://github.com/effekt-lang/effekt/pull/527#issuecomment-2253022331

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:132]: This is the same as:

```suggestion
def limit[A, R](number: Int) { stream: () => R / Emit[A] }: R / { Emit[A], Done } = {
  var i = 0;
  for { stream() } {
    if (c < number) { i = c + 1; do emit(v) }
    else { do done() }
  }
}
```

isn't it?

Comment (b-studios) [reply libraries/common/stream.effekt:132]: Also `limit` seems like it would limit the emitted values, not their number. Why not the standard `take`?

**Comment by b-studios** [comment]
Can we maybe translate some Unison examples to Effekt to compare? That would be nice. @phischu could you please look into the Unison implementation as well?

File: libraries/common/stream.effekt
Comment (b-studios) [reply libraries/common/stream.effekt:156]: Also: I really, really, really, would prefer only returning a list and not a pair.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:188]: This is buggy and needs to be commented out to try the library at all.

The error I get is:

```
Expected Tuple2[R, Array[A]] but got Unit.
```

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:234]: This also doesn't work at all and results in a lot of type errors

**Comment by b-studios** [comment]
If you comment out the broken functions you can actually write an example which might look like this:

```
  val (_, l) = gatherList[Int, Unit] {
    for[Int, Unit] { range(0, 5) } { i =>
      for[Int, Unit] { range(1, 3) } { j =>
        println(i + j)
      }
    }
  }

  println(l)
  ```
  
  Please note the explicit type arguments which are necessary because I need to say WHICH stream I want to handle.

File: libraries/common/stream.effekt
Comment (phischu) [reply libraries/common/stream.effekt:16]: If we panic, we have to call it `unsafeRepeat`. I'm doing nothing for now. I wish we had `Nat`.

File: libraries/common/stream.effekt
Comment (phischu) [reply libraries/common/stream.effekt:156]: > That is, if this is actually helpful in our current implementation.
Is this about performance? We can benchmark later.

File: libraries/common/stream.effekt
Comment (phischu) [reply libraries/common/stream.effekt:77]: Let's leave this open until we have more experience

**Comment by phischu** [comment]
> Please note the explicit type arguments which are necessary because I need to say WHICH stream I want to handle.

How about

```
on[Int].for { ...
```

Or we try to fix this at the language level.

**Comment by phischu** [comment]
> (lazy nondeterminism from a stream)

This is planned in Phase 3

**Comment by phischu** [comment]
> could you please look into the Unison implementation as well?

I am looking at the [`streaming`](https://hackage.haskell.org/package/streaming) library, which is great.

**Comment by b-studios** [comment]
on the language level probably means fixing type inference which is planned but will take a while

File: libraries/common/bytearray.effekt
Comment (b-studios) [inline libraries/common/bytearray.effekt:79]: `into` sounds like the string would be mutable.

File: examples/benchmarks/input_output/word_count_ascii.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_ascii.effekt:25]: Honestly, I would annotate `read` to disambiguate more locally
```suggestion
  exhaustively { do read[Byte]() } { c =>
```

File: examples/benchmarks/input_output/word_count_ascii.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_ascii.effekt:52]:
```suggestion
    for { range(32, 127) } { c =>
```

not necessary anymore.

File: examples/benchmarks/input_output/word_count_ascii.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_ascii.effekt:47]: Does this work on windows machines?

File: examples/benchmarks/input_output/word_count_ascii.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_ascii.effekt:68]: Fantastic case study, thanks!

File: examples/benchmarks/input_output/word_count_utf8.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_utf8.effekt:14]: See above, we use it a second time here, maybe move?

File: examples/benchmarks/input_output/word_count_utf8.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_utf8.effekt:25]: Same as above.

File: examples/benchmarks/input_output/word_count_utf8.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_utf8.effekt:53]:
```suggestion
    for { range(128512, 128522) } { c =>
```

File: examples/benchmarks/input_output/word_count_utf8.effekt
Comment (b-studios) [inline examples/benchmarks/input_output/word_count_utf8.effekt:70]: Again, great example.

File: libraries/common/array.effekt
Comment (b-studios) [inline libraries/common/array.effekt:128]: Minor: we should be starting to use `///`

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:11]: Beautiful!

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:21]: `continually` is a strange word to remember and type.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:64]: What does this mean? You don't want to support it, or you don't yet?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:67]: We should have some common best practice on where to define effects. In the middle of files, all at the top?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:77]: Again, is there a better word than `exhaustively`?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:107]: Maybe `indexed`?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:122]: If number is 0, do we even need to run `stream`?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:135]: Not consistent with the implementation above. Also please use `///`

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:156]: You know I am not a big fan of returning two things here. Same below.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:187]: Should probably be called `collectByteArray`.

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:194]: Is this now the best resizing that we agreed on?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:206]: Again, as above: do we want to sprinkle effect definitions all over files?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:270]:
```suggestion
  } with read[A] {
```

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:283]: It would be cool to have a separate standard library that helps with file sizes and other units, like
```
1.MB
100.ms
etc.
```

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:299]: Why do you use the long form here?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:330]:
```suggestion
  } with read[Byte] {
```

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:308]: That's really cool!

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:360]: Why do you mix `do read().toInt` and the decoding? Shouldn't the decoding go into strings or char or whatever?

File: libraries/common/stream.effekt
Comment (b-studios) [inline libraries/common/stream.effekt:379]: Same as above.

File: libraries/common/stream.effekt
Comment (omentic) [reply libraries/common/stream.effekt:156]: I agree... if I understand the design of having an overt generic `R` and supporting streams that step to it (that then, you get streams stepping to `Unit` for-free) then I think it necessitates duplicating all these methods here that want to *do* something with the overt `R`. Since when `stream` is `() => Unit / emit[A]` (probably the most common use case?) you're always getting back this empty tuple `((), [...])` instead of just `[...]`

I don't really see the use case for having an overt generic `R` to begin with, to be honest. A Stream seems like something that should always step to the unit type. But I also haven't really worked with explicit Stream APIs all that much.

It might be worth considering having an alias `Stream` to `() => Unit` and a `Stream[R]` to `() => R`? And implementing separate functions for `Stream`, which mostly are direct wrappers around the corresponding functions for `Stream[R]`, but discard the empty tuple in these cases?

File: libraries/common/stream.effekt
Comment (omentic) [inline libraries/common/stream.effekt:9]: I would personally prefer this to be Uppercase for consistency with the other effects I've come across...

**Review by omentic — COMMENTED** [review]
This is really cool! I gave some surface-level feedback. I implemented a kinda similar API recently and will go through and give deeper feedback when less busy (if this hasn't been merged by then :smile:)

File: libraries/common/stream.effekt
Comment (b-studios) [reply libraries/common/stream.effekt:77]: Please explain what this function does

File: effekt/jvm/src/test/scala/effekt/ChezSchemeTests.scala
Comment (phischu) [reply effekt/jvm/src/test/scala/effekt/ChezSchemeTests.scala:30]: Perhaps, but right now extern definitions are missing.

File: effekt/jvm/src/test/scala/effekt/LLVMTests.scala
Comment (phischu) [reply effekt/jvm/src/test/scala/effekt/LLVMTests.scala]: There is a separate issue that should be resolved separately.

File: libraries/common/stream.effekt
Comment (b-studios) [reply libraries/common/stream.effekt:350]: We talked about this already... it can't be do to the character wise streaming `do read()`

File: effekt/jvm/src/test/scala/effekt/LLVMTests.scala
Comment (b-studios) [inline effekt/jvm/src/test/scala/effekt/LLVMTests.scala:34]: Is there an issue already? It would be horrible if this can't be run on LLVM only due to some bug.

Comment (phischu) [reply effekt/jvm/src/test/scala/effekt/LLVMTests.scala:34]: No, but I am too lazy to minimize these. We can always fix this later.

Comment (b-studios) [reply effekt/jvm/src/test/scala/effekt/LLVMTests.scala:34]: Here is the function type that is failing when instantiated:

```
Function(List(),List(),List(),List(Interface(stop,List())),Data(Byte,List()))
```
which is the inferred type of

```
{ (){g7113} =>
  shift(p7114) { (){k7115} =>
    def resume3708(){h7116} = resume(k7115) { h7116(g7113) }

    resume3708({ (){stop4511} =>
      val v_r_49857117: Int398 = !i3707;
      if (infixLt175(v_r_49857117, size2454(bytes3621))) {
        val v_r_49867118: Int398 = !i3707;
        let v_r_49877119 = unsafeGet2457(bytes3621, v_r_49867118)

        v_r_49877119
      } else {
        stop4511.stop()
      }
    })
  }
}
```

Comment (b-studios) [reply effekt/jvm/src/test/scala/effekt/LLVMTests.scala:34]: This is the function which is causing the problem:

```
def feed[T](array: Array[T]) { reader: () => Unit / read[T] } = {
  var i = 0
  try {
    reader()
  } with read[T] {
    resume {
      if (i < array.size) {
        array.unsafeGet(i)
      } else {
        do stop()
      }
    }
  }
}
```

Comment (b-studios) [reply effekt/jvm/src/test/scala/effekt/LLVMTests.scala]: Fixed by f30d83a. I just forgot to bind `cps`...

**Comment by b-studios** [comment]
@phischu I took the liberty to add some doc comments.

**Comment by b-studios** [comment]
Interesting, the tests pass on my machine but apparently they now leak memory...

**Comment by b-studios** [comment]
To avoid having to rebase over and over again, I am in favor of just merging this and revising later. It is in a state that is already useful and we can continue improving it.

@phischu I'll give you the approve and if you deem it ready, just merge.

**Review by b-studios — APPROVED** [review]

**Comment by jiribenes** [comment]
Before merging, please rename the PR (just so that the autogenerated Changelog for the next release is a little more understandable 😇)

**Pull request merged**
